### ISSUE TITLE
[Feature Request] ASHA

### ISSUE THREAD

**Issue by bbudescu** [automl/SMAC3#1169]
Asynchronous Successive Halving Algorithm (ASHA) is, as far as I can understand, a tweaked version of the Hyperband multi-fidelity intenstifier (aka [ early stopping | pruning | scheduler ] algorithm) that allows to make better use of the available computing power in the context of many cpu cores (improves cpu occupancy).

Here are a bunch of references:
- nice [explanation](https://d2l.ai/chapter_hyperparameter-optimization/sh-async.html) in a book that I found helpful for myself to understand how it works
- authors' 2018 [blog post](https://blog.ml.cmu.edu/2018/12/12/massively-parallel-hyperparameter-optimization/) introducing the method
- 2018 [paper](https://openreview.net/revisions?id=S1Y7OOlRZ) (rejected by ICLR); arxiv preprint [here](https://arxiv.org/abs/1810.05934)
- another docs entry offering a pretty nice [explanation](https://docs.determined.ai/latest/model-dev-guide/hyperparameter/search-methods/hp-adaptive-asha.html)

I think [here](https://github.com/liamcli/darts_asha) is the original repo in which the authors published the reference implementation, but a bunch of other optimizers implement it, too nowadays (e.g., [Oríon](https://orion.readthedocs.io/en/stable/user/algorithms.html#asha), [optuna](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.SuccessiveHalvingPruner.html)). I've personally used `ray[tune]`'s implementation before (docs [overview](https://docs.ray.io/en/latest/tune/api/schedulers.html#asha-tune-schedulers-ashascheduler), [api reference](https://docs.ray.io/en/latest/tune/api/doc/ray.tune.schedulers.AsyncHyperBandScheduler.html#ray.tune.schedulers.AsyncHyperBandScheduler)), and, at some point, I somehow ended up stepping through the [code](https://github.com/ray-project/ray/blob/f1316e8af8c8930cd062f78b1d46f7933f744bf8/python/ray/tune/schedulers/async_hyperband.py#L19) and their implementation looked surprisingly simple (yes, I know... famous last words, but after reading research papers there's sometimes this shock when you read the code and see that the implementation could have been explained in much simpler words).

**Motivation:** I was checking out an optimization session running on 64 cores and noticed the cpu load drop below 50% after 12 hours. Not sure yet exactly how many trials it had gotten through at that point (after 22 hours it got through 12600 trials, and cpu load is under 40%).

Also, not sure if this is really the bottleneck that throttles cpu usage, I know it can be something else, too, like, e.g., many processes waiting for the surrogate model to retrain before starting a new trial or something (that's why I also mentioned the number of finished trials above) or perhaps other reasons, and it's kinda hard to debug, so I thought it'd be useful to be able to eliminate some potential causes, and this is the one I thought of first.

However, if you have some suggestions on improving parallelism, please let me know.

**Comment by eddiebergman** [comment]
As an additional point, ASHA is not model based, and so the slow-down you're experiencing if you're using a model based optimizer is that as the number of trials increases, the model re-training takes longer and longer. One of the benefits of using SMAC's random forest surrogate is that it is generally faster than re-training the de-facto standard of a GP, which is O(n^2) complexity in the number of trials.

ASHA in comparison is (usually) fully non-model based and so the speed ups in consideration there are largely at a cost of optimization taking longer to converge to some optimal solution.

Still fully onboard with an ASHA implementation in SMAC, just that better core utilization does not directly translate to better end performance, and perhaps running on less cores with a better optimization routine is still something worth considering.

**Comment by bbudescu** [comment]
Thanks for your answer. I'm not sure I understand fully correct, but just to make sure, I don't mean to use ASHA as a standalone optimizer, but in conjunction with SMAC's random forrest surrogate model, as a multi-fidelity intensifier in place of Hyperband. So just to cut short trials suggested by the RF that underperform.

**Comment by bbudescu** [comment]
Also, I guess, in my case, to eliminate the other probable cause for the low cpu load, I think it would make sense to increase `retrain_after`, as suggested in the [docs](https://automl.github.io/SMAC3/main/advanced_usage/12_optimizations.html).

However, could you suggest a particular value given the metrics above? More generally, what'd be a good way to compute optimal value for `retrain_after` given the average time it takes to run a trial and number of workers in order to keep cpu load above a certain threshold of, e.g., 90%? 

Also, should the hyperband reduction factor play a role in establishing the optimal value for `retrain_after`?

**Comment by bbudescu** [comment]
Another nice thing one could do if there's a bottleneck like this would be to employ an async polling method, i.e., just use an older version of the surrogate model to provide trial configs when asked for while the new surrogate model is training. 

I guess there's an optimal tradeoff to be calculated here, too - figuring out when it is better to actually wait for a little while for the new results to be churned by the surrogate model before starting a new time-consuming trial, and when is it better to just keep starting almost-random trials (i.e., because too little data to constrain the model) instead of remaining blocked.

And I think that's a bit meta right there, because I guess that this would mean the decision of asking the old model now vs waiting for the new one to finish training would, fundamentally, be based on the expected improvement of the accuracy of the surrogate model. 

And now my head hurts.

**Comment by bbudescu** [comment]
Sorry I over-complicated things. It can be simpler than all that estimation. Just use the old model if occupancy falls below 90%. Occupancy is the percentage of workers that are running, i.e., that aren't waiting for a config suggestion from the surrogate model.

**Comment by mwever** [comment]
Hello @bbudescu 

Do I understand the thread correctly that you would still like to request ASHA as a new feature while the rest has been transferred to the new ticket with ID #1170?

**Comment by bbudescu** [comment]
Hi @mwever,

> Do I understand the thread correctly that you would still like to request ASHA as a new feature while the rest has been transferred to the new ticket with ID https://github.com/automl/SMAC3/issues/1170?

Yes, that's the main idea.

**Comment by mwever** [comment]
Alright, so right now we do not have the capacity to add ASHA as a new feature but we would highly appreciate your contribution in the form of a PR

**Comment by bbudescu** [comment]
> right now we do not have the capacity to add ASHA as a new feature

I totally get that. I can see you're a small team, but you've been doing a great job. Thanks for your work and dedication!

>  we would highly appreciate your contribution in the form of a PR

I'll do it when I have some time again, but in the mean time, can you, please offer some support for merging https://github.com/automl/SMAC3/pull/1178, which is supposed to address https://github.com/automl/SMAC3/issues/1170?

Currently, I'm in a testing stage where I mean to run another 48h optimization session on 64 cores with the new background training to see how CPU load and results are affected, to compare them against the previous session.
